### PR TITLE
cortexm: fix thread_yield() -> thread_yield_higher() in ISRs

### PR DESCRIPTION
**Pull request by kaspar030**
Threads with the same priority are supposed to be scheduled cooperatively, but most ISR's call ```thread_yield()```, causing the current priority's circular list to advance.

Automatically done using

```git grep -l sched_context_switch_request | xargs perl -0pi -e 's/if \(sched_context_switch_request\) \{\n        thread_yield\(\);/if \(sched_context_switch_request\) \{\n        thread_yield_higher\(\);/'```

### CODE REVIEW

**Comment by kaspar030** [comment]
Marked WIP as I'd like to create a simple reproducer showing the problem, fix the docs and ... do some superficial testing. :)

**Comment by kaspar030** [comment]
Hm, this should be a function:

```
static inline void sched_isr_exit(void)
{
    if (sched_context_switch_request) {
        thread_yield_higher();
    }   
}
```

... thing is, it doesn't really belong into thread.h, but as the includes go, it depends on thread_yield_higher(), which is defined in thread.h.

Maybe we should rename ```thread_yield*()``` into ```sched_yield*()```?

**Comment by spiderkeys** [comment]
Thanks for looking into this. Was an interesting case in my project particularly because my application is structured to use a number of threads at the same priority and ISRs could plausibly starve some threads by yielding them to their siblings constantly.

Concerning where you could place sched_isr_exit, how about core/arch/thread_arch? This would give you the flexibility to implement it for each platform depending on differences in ISR handling and you already have your thread.h dependencies there. Maybe a sched_arch file is in order for consistency. Short of using macros, I don't see a way of getting it inlined into a core header file without sacrificing arch specific flexibility.

Taking a look at the UART ISR for the msp430, it looks like it behaving strangely as well in that it calls sched_run instead of thread_yield or thread_yield_higher inside of __exit_isr. At first glance, this looks like it will not schedule a higher priority thread to run if it was signaled during the ISR, so while it solves the issue with sibling threads, it looks like it doesn't work when there is a request for a context switch to a higher priority thread.

If I understand that correctly (and I may well be missing something), then creating a specific sched_isr_exit method using thread_yield_higher seems like the right course of action, coupled with a bit more documentation for CPU platform developers to understand the mechanisms involved when writing their ISRs.

**Comment by haukepetersen** [comment]
> Hm, this should be a function:

done: #6171

**Comment by lebrush** [comment]
@kaspar030 please update after #6171

**Comment by aabadie** [comment]
@kaspar030, do you think this one could be part of the next release ?

**Comment by kaspar030** [comment]
- rebased
- adapted to #6171

**Comment by kaspar030** [comment]
This is ready for review. testing could involve reverting the first commit, after which the supplied test application should fail...

**Comment by kaspar030** [comment]
(accidentaly squashed blacklisting of nucleo32-f031)

**Comment by kaspar030** [comment]
*ping* @haukepetersen wanna take a look?

**Comment by smlng** [comment]
test succeed with and w/o the fix, did some other changes fixed that already? tested with PhyNode

**Comment by kaspar030** [comment]
> test succeed with and w/o the fix, did some other changes fixed that already? tested with PhyNode

Weird. Anyhow, I fixed the test, it should now work as expected (fail without fix, work with fix).

**Comment by kaspar030** [comment]
@smlng could you try again?

**Comment by kaspar030** [comment]
*ping*

@kYc0o maybe you could take a look?

**Comment by kaspar030** [comment]
@aabadie flake8 complains on the test [here](https://ci.riot-os.org/RIOT-OS/RIOT/6164/b4ecf0d09ef7cc678c89f5b46dcbd29d2fe89750/output/static_tests.txt). How do I fix the "module import not at top of file"? I think the python path needs to be above, right?

File: tests/isr_yield_higher/tests/test.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (C) 2017 Kaspar Schleiser <kaspar@schleiser.de>
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+import os
+import sys
+
+sys.path.append(os.path.join(os.environ['RIOTBASE'], 'dist/tools/testrunner'))
```
Comment (aabadie) [inline tests/isr_yield_higher/tests/test.py]: Move those 2 lines in the `main` block to fix the flake8 issue:

```python
if __name__ == "__main__":
    sys.path.append(os.path.join(os.environ['RIOTBASE'], 'dist/tools/testrunner'))
    from testrunner import run
    sys.exit(run(testfunc))
```

Comment (kaspar030) [reply tests/isr_yield_higher/tests/test.py]: thanks for the fast response! fixed

**Pull request merged**